lib/model/indexhandler.go

```diff
@@ -328,12 +348,42 @@ func (s *indexHandler) receive(fs []protocol.FileInfo, update bool, op string) e
 	if !update {
 		fset.Drop(deviceID)
 	}
+
+	l.Debugf("Received %d files for %s from %s, prevSeq=%d, lastSeq=%d", len(fs), s.folder, deviceID.Short(), prevSequence, lastSequence)
+
+	// Verify that the previous sequence number matches what we expected
+	if prevSequence > 0 && prevSequence != fset.Sequence(deviceID) {
+		l.Warnf("Bug: device %v folder %s sent index update with sequence %d while we expected %d", deviceID.Short(), s.folder, fset.Sequence(deviceID), s.prevSequence)
+		s.evLogger.Log(events.Failure, "index update with unexpected sequence")
+	}
+
 	for i := range fs {
+		// Verify index in relation to the claimed sequence boundaries
+		if fs[i].Sequence < prevSequence {
+			l.Warnf("Bug: device %v folder %s sent file with sequence %d outside of claimed range %d-%d", deviceID.Short(), s.folder, fs[i].Sequence, prevSequence, lastSequence)
+			s.evLogger.Log(events.Failure, "file with sequence before prevSequence")
+		}
+		if lastSequence > 0 && fs[i].Sequence > lastSequence {
+			l.Warnf("Bug: device %v folder %s sent file with sequence %d outside of claimed range %d-%d", deviceID.Short(), s.folder, fs[i].Sequence, prevSequence, lastSequence)
+			s.evLogger.Log(events.Failure, "file with sequence after lastSequence")
+		}
+		if i > 0 && fs[i].Sequence <= fs[i-1].Sequence {
+			l.Warnf("Bug: device %v folder %s sent non-increasing sequence %d after %d", deviceID.Short(), s.folder, fs[i].Sequence, fs[i-1].Sequence)
+			s.evLogger.Log(events.Failure, "non-increasing sequence")
+		}
```

Comment:
These look largely "symmetric" to existing checks in `sendIndexTo`. Lets use the same pattern there as well:
That panic behind `shouldDebug` could also become an unconditional check with failure-logging intead of panicking. And also sending a failure report when detecting a non-increasing sequence there.